shared-services/README.md

```diff
@@ -0,0 +1,172 @@
+    ];
+    accept-flake-config = true;
+    allow-import-from-derivation = true;
+  };
+}
+```
+You may edit this file in three ways:
+1. Fill the `description` to your liking.
+2. Edit the `nixConfig` attrset, but do not **no** invalidate the cache-related fields.
+3. Add new inputs to the flake. *Note that `devx` provisions and manages all other flake inputs, and therefore should be the only input in your flake. You may nonetheless declare additional inputs as you would normally, and they will be available in `devx.nix`.*
+
+# The `devx.nix` file
+
+The `devx.nix` file is the entrypoint to all your nix code:
+```nix
+devx: 
+{
+  supportedSystems = ["x86_64-linux" "x86_64-darwin"]; 
+  
+  ghcVersions = ["ghc8107" "ghc901" "ghc924"]; 
+
+  haskellNixProject = ./haskell.nix;
+
+  extraFlakeOutputs = {};
+
+  shellHooks = {
+    default = {
+      hook1-name = ''bash code'';
+      hook2-name = ''bash code'';
+    };
+    shell1 = {};
+    shell2 = {};
+  };
+
+  shellCommands = {
+    default = {
+      command1-name = {
+        description = "";
+        category = "";
+        exec = "bash code";
+      };
+      command2-name = {
+        description = "";
+        category = "";
+        exec = "bash code";
+      };
+    }];
+    shell1 = {};
+    shell2 = {};
+  };
+}
+```
+    
+- `supportedSystems` should be any of "x86_64-linux" "x86_64-darwin" "aarch64-linux" "aarch64-darwin".
+  Flake outputs for packages and devshells will be produced for each supported system.
+    
+- `ghcVersions` should be any of "ghc8106" "ghc901" "ghc924". This will determine the version of the haskell toolchain. A devShell will be avaiable for each GHC version.
+
+- `haskellNixProject` path to a standard `./haskell.nix` file. Click [here](#the-haskellnix-file) for reference. 
+
+- `extraFlakeOutputs` should have the same format as standard flake outputs. The outputs will be merged with the outputs generated by `devx` (more on this below).
+
+- `shellHooks.{shellName}` is an attrs where each key is the hook name and each value is bash code to be executed each time that devShell is entered.
+
+- `shellCommands.{shellName}` is an attrs where each key is the command name and each value is an attrs defining the command. A menu of available commands is displayed upon entering that devShell. Commands are grouped by `category`.
+
+# The `devx` attribute set 
+
+The `devx` attribute set contains everything that you'll need in your nix code.
+```nix
+devx = {
+  self = {}; 
+  inputs = {};
+  outputs = {};
+  lib = {
+    mkFlakeOutputs = {};
+  };
+}
+```
+
+- `self`: the standard "self" flake input.
+
+- `inputs`: the desystemized inputs coming from `flake.nix` plus the desystemized implicit inputs. The implicit inputs are: 
+  - `nixpkgs`
+  - `haskell-nix`
+  - `hackage-nix`
+  - `CHaP`
+  - `iohk-nix`
+  - TODO unexhaustive list
+
+- `outputs`: the final outputs coming from `devx.nix`'s `extraFlakeOutputs` plus the implicit outputs. The implicit outputs are:
```

Comment:
We should also have standalone documented functions corresponding to these that are orthogonal to how the flake is set up. E.g. some function `makeHaskellPackages` that takes the cabal.project and relevant config and returns a `ghcVersion -> packages` attrset.